examples/research_projects/jax-projects/README.md

```diff
@@ -383,13 +383,212 @@ official [flax example folder](https://github.com/huggingface/transformers/tree/
+
+```python
+class ModelFlax:
+
+  def __init__(self, config):
+    self.config = config
+    self.key_proj = flax.linen.Dense(config)
+    self.value_proj = flax.linen.Dense(config)
+    self.query_proj = flax.linen.Dense(config)
+    self.logits_proj = flax.linen.Dense(config)
+```
+
+At first glance the linear layer class `flax.linen.Dense` looks very similar to PyTorch's `torch.nn.Linear` class. However, instantiating an object `model_flax` only defines the linear transformation functions and does **not** allocate memory to store the linear transformation weights. In a way, the attribute `self.key_proj` tell the instantiated object `model_flax` to perform a linear transformation on some input and force it to expect a weight, called `key_proj`, as an input.
+
+This time we would illustrate the object `model_flax` without the weight matrices:
+
+![alt text](https://raw.githubusercontent.com/patrickvonplaten/scientific_images/master/lm_flax_def.png)
+
+
+Accordingly, the forward pass requires both `input_ids` as well as a dictionary consisting of the model's weights (called `state` here) to compute the `sequences`:
+
+To get the initial `state` we need to explicitly do a forward pass by passing a dummy input:
+
+```python
+state = model_flax.init(rng, dummy_input_ids)
+```
+
+and then we can do the forward pass.
+
+```python
+sequences = model_flax.apply(input_ids, state)
+```
+
+Visually, the forward pass would now be represented as passing all tensors required for the computation to the model's object:
+
+![alt text](https://raw.githubusercontent.com/patrickvonplaten/scientific_images/master/lm_flax_inference.png)
+
+This design is called **stateless** because the output logits, the `sequences`, **cannot** change if the word embeddings, the `input_ids`, stay the same. Hence, the function's output only depends on its inputs, being the `input_ids` and the `state` dictionary consisting of the weights **state**, `[key_proj, value_proj, query_proj, logits_proj]`. 
+
+Another term which is often used to describe the design difference between Flax/JAX and PyTorch is **immutable** vs **mutable**. A instantiated Flax model, `model_flax`, is **immutable** as a logical consequence of `model_flax`'s output being fully defined by its input: If calling `model_flax` could mutate `model_flax`, then calling `model_flax` twice with the same inputs could lead to different results which would violate the "*statelessness*" of Flax models.
+
+Now let us see how this is handled in `Transformers.` If you have used a flax model in Transformers already, you might wonder how come we don't call `model.init` and `model.apply`. This is because the `FlaxPreTrainedModel` class abstracts it away. 
```

Comment:
```suggestion
Now let us see how this is handled in `Transformers.` If you have used a Flax model in Transformers already, you might wonder how come we don't call `model.init` and `model.apply`. This is because the `FlaxPreTrainedModel` class abstracts it away. 
```

I like this, but we don't really talk about `.apply()` and `.init()` before this paragraph. So maybe we could just write instead "you might wonder how come you don't always have to pass the parameters to the function of the forward pass."